src/utils/hooks/use-validation-schema.ts

```diff
@@ -1,45 +1,81 @@
+import isEmpty from "lodash/isEmpty";
 import { useContext, useEffect, useState } from "react";
 import * as Yup from "yup";
 import { ObjectShape } from "yup/lib/object";
-import { IYupValidationRule, YupContext, YupHelper } from "../../components/frontend-engine/yup";
+import { TFrontendEngineValues } from "../../components/frontend-engine/types";
+import { IFieldYupConfig, TFormYupConfig, YupContext, YupHelper } from "../../components/frontend-engine/yup";
+import { ObjectHelper } from "../object-helper";
 
 /**
- * Hook that interacts with the validation schema context provider
- * use this hook to get/set the validationSchema and config
+ * Hook that handles the generation of the validationSchema
  */
 export const useValidationSchema = () => {
-	const { formValidationConfig, setFormValidationConfig } = useContext(YupContext);
-	const [validationSchema, setValidationSchema] = useState<Yup.ObjectSchema<ObjectShape>>();
+	const { formValidationConfig } = useContext(YupContext);
+	const [hardValidationSchema, setHardValidationSchema] = useState<Yup.ObjectSchema<ObjectShape>>();
+	const [softValidationSchema, setSoftValidationSchema] = useState<Yup.ObjectSchema<ObjectShape>>();
+	const [warnings, setWarnings] = useState<Record<string, string>>({});
 
 	useEffect(() => {
 		if (formValidationConfig) {
-			setValidationSchema(YupHelper.buildSchema(formValidationConfig));
+			let hardValidationConfig: TFormYupConfig = {};
+			let softValidationConfig: TFormYupConfig = {};
+
+			Object.entries(formValidationConfig).forEach(([key, value]) => {
+				if (!value.validationRules || value.validationRules.length === 0) {
+					// NOTE: Fallback to default yup validators that's being set in component level
+					hardValidationConfig = ObjectHelper.upsert(hardValidationConfig, key, {
+						schema: value.schema,
+						validationRules: [],
+					});
+					return;
+				}
+
+				value.validationRules.forEach((rule) => {
+					const config = {
+						schema: value.schema,
+						validationRules: value.validationRules ? [...value.validationRules, rule] : [rule],
+					};
+					if (rule.soft) {
+						softValidationConfig = ObjectHelper.upsert<IFieldYupConfig>(softValidationConfig, key, config);
+					} else {
+						hardValidationConfig = ObjectHelper.upsert<IFieldYupConfig>(hardValidationConfig, key, config);
+					}
+				});
```

Comment:
it appears that `validationRules: value.validationRules ? [...value.validationRules, rule] : [rule]` is causing the duplication, I have decided to remove the forEach loop.

now it will filter the soft / hard validation rules and upsert to the configs depending whether the respective rules exist. this logic will need to be changed further though, I think it can simplified further when we integrate the validation-schema-generator.